### ISSUE TITLE
Create XML schema for roslaunch files

### ISSUE THREAD

**Issue by jbohren** [ros/ros_comm#455]
@Fran-B After you're done with the package.xml schema over in ros-infrastructure/rep/issues/77, do you have any interest in helping define a schema for roslaunch?

The roslaunch sources are here in this package: https://github.com/ros/ros_comm/tree/indigo-devel/tools/roslaunch

**Comment by Fran-B** [comment]
Jonathan, I don't know what needs to be done there.
Can you point me at an XML file, and/or documentation for same, that you would want a schema for?

**Comment by jbohren** [comment]
> Jonathan, I don't know what needs to be done there.
> Can you point me at an XML file, and/or documentation for same, that you would want a schema for?

The current documentation for roslaunch XML is here: http://wiki.ros.org/roslaunch/XML

**Comment by Fran-B** [comment]
Now I see the need.  I will see what I can do to help out here.
I would also appreciate learning to understand these files at the same time.
Are the element nodes inside "launch" required in any order, or are multiple instances of them allowed intermingled in any order?

**Comment by jbohren** [comment]
There's no ordering requirement.

**Comment by nalt** [comment]
Has there been any progress on this? Would DTD or XSD be better?

**Comment by dirk-thomas** [comment]
> Has there been any progress on this?

Since there is no update on this ticket for 3+ years you can assume that there has been no progress on this.

> Would DTD or XSD be better?

I would suggest XSD which is the more modern and more capable format. We also use it for the package manifest files.

**Comment by Fran-B** [comment]
I thought that I completed this task 3 years ago.  I supplied a completed XSD file to those involved to integrate into source and create a pull request.
Since then something went wrong with my connection to ROS answers and I no longer receive email from the list.

I am sure that I can find the completed file and email it to somebody for integration into a pull-request.
I myself would not know where it should be included within the source tree, and ROS has changed versions since then any way.

I put a lot of work into this and would like it to be added into the system.

Looking back at the original email chain Jack O'Quin said that he forwarded the completed schema file to Dirk on July 7, 2014

Bottom line is that we just need somebody to integrate the file into the tool chain and create a pull request.  I would have no clue how to do either task and have never had a copy of the ROS source tree.

Would either of you please find somebody to perform this last step top complete this task?

Thanks,
Fran

**Comment by dirk-thomas** [comment]
@Fran-B I just checked the emails and the schema file you emailed was for the package manifest - not for launch files.

**Comment by nalt** [comment]
I created an XSD (and also DTD) and put it on https://gist.github.com/nalt/dfa2abc9d2e3ae4feb82ca5608090387 for now.

Code completion, syntax highlighting and validation work well, except for the cases mentioned there.

Please feel free to comment.

**Comment by mikepurvis** [comment]
@nalt That looks great! I tried validating all our launch files with this just to see how they fared:

```
find /path/to/ros -type f -name "*.launch" -exec xmllint -schema roslaunch.xsd {} --noout \;
```

Some bugs I think this found:

- `group` nested in `group` shouldn't throw an error.
- True/true/False/false should all be acceptable for `respawn`.
- Anywhere an enumeration (or really any attribute) is expected, a `$(.*)` needs to also be acceptable.
- `<launch deprecated="[msg]">` needs to be permitted.

It would be great to be able to validate other stuff like the validity of yaml inside `<rosparam>` tags and the format of `$()` expressions, but I can appreciate that we're constrained in a lot of this by the capabilities of the definition languages— past a certain point is the role of `roslaunch-check`.

**Comment by Fran-B** [comment]
Sorry Dirk, I guess that I am confused why I was associated with this issue then.

Sorry for the confusion.

**Comment by nalt** [comment]
About @mikepurvis points:

- Nesting groups is not mentioned in the Wiki, but I have also often seen it.

- The wiki never mentions True/False with a capital letter. Shouldn't it rather be rejected?
 The Wiki even doesn't allow false for respawn and some other attributes, but I guess this is not intended.

**Comment by nalt** [comment]
I updated the XSD. It addresses all issues mentioned by @mikepurvis, except for True/False. The allowable commands in `$(` are also checked.

**Comment by l3s2d** [comment]
Any updates on this? The schema here seems to work: https://gist.github.com/nalt/dfa2abc9d2e3ae4feb82ca5608090387

It would be great if it was available at a ros.org url, similar to the package.xml schemas.

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
